### ISSUE TITLE
Formula-specific lesson migration for Modern AI I–VIII

### ISSUE THREAD

**Issue by JJo-0** [JJo-0/JJo-0.github.io#38]
## Problem

The current `쉽게 설명 + 계산 과정` cards reuse a small set of generic visual families. That makes many visuals mathematically unrelated to the exact equation being explained. Formula count parity is not a sufficient quality metric.

## Target contract

Every display equation must have an explicit `FormulaLessonRef` keyed by formula ID and one of these states:

- `approved-interactive`
- `approved-derivation`
- `approved-structure`
- `no-visual-with-reason`
- `unreviewed` (migration-only; never renders a visual)

No visual may be selected directly from a TeX keyword heuristic in production. The exact equation ID selects the lesson group, focus step, renderer, source status, assumptions, and no-visual rationale.

## Architecture

```text
formulaId
  -> FormulaLessonRef
  -> FormulaLessonSpec
  -> pure math model
  -> synchronized equation / table / visualization / checks
```

Several consecutive equations may share one lesson group. For example, Part II `MAI-P2-020`–`MAI-P2-027` form one least-squares / normal-equation derivation with a shared `X`, `y`, and `w`; opening a particular equation highlights its corresponding step.

### Phase 0 — stop misleading visuals

- [ ] Remove the generic `y=x²` fallback.
- [ ] Stop rendering a visual for unreviewed formulas.
- [ ] Separate unrelated families such as dimension growth and K-fold.
- [ ] Generate a 303-equation migration inventory from the protected Part I and Part II ledgers.
- [ ] Add fail-closed checks: no heuristic renderer in production, no unreviewed formula may render a visual, and every display formula must be in the inventory.

### Phase 1 — Part II golden implementation

- [ ] Implement `MAI-P2-017` dot-product prediction.
- [ ] Implement `MAI-P2-020`–`MAI-P2-027` least-squares / normal-equation lesson.
- [ ] Implement `MAI-P2-028`–`MAI-P2-033` matrix derivative and outer-product accumulation.
- [ ] Implement generalization, regularization, K-fold, SVM/kernel, minimum-distance, Bayes posterior/risk, Gaussian QDA/LDA, Naive Bayes, and dimension growth.

### Phase 2 — Part I linear algebra

Build reusable primitives for vector arithmetic, dot products, matrix transforms, span/rank, norms, determinant, inverse/singularity, quadratic forms, PSD/PD, eigenvectors, covariance, and PCA.

### Phase 3 — Part I probability and optimization

Build expectation, variance, conditional probability, Gaussian, loss surface, gradient field, and step-size stability lessons.

### Phase 4 — Parts III–VIII

Apply the same exact-ID lesson contract to perceptron/MLP/CNN, optimization, image classification, segmentation, denoising/VAE/diffusion, and contrastive learning.

## Acceptance criteria

- Formula TeX and SHA-256 source contracts remain unchanged.
- Visual controls correspond to symbols that actually occur in the equation or its explicitly linked derivation group.
- The equation, numeric table, visualization, and checks all use the same pure computation result.
- Source-suspect and corrected variants remain visibly distinct.
- Lazy loading remains formula-local and mobile-safe.
- Unit tests validate numerical outputs and invariants such as posterior sum = 1, MSE >= 0, covariance symmetry, and matrix dimensions.

**Comment by JJo-0** [comment]
## 2026-08-19 migration checkpoint — draft PR #40

A continuing draft PR is now open: #40 (`agent/formula-specific-lesson-migration`).

### Completed

- Replaced generic family-based visual selection with an exact formula-ID registry.
- Removed the generic `y=x²` fallback and retired `FormulaVisual` runtime.
- Inventoried all Part I–II display formulas: 303 total (Part I 238, Part II 65).
- Unreviewed formulas now render no visual rather than a mathematically unrelated placeholder.
- Added lazy dynamic renderer loading and a fail-closed formula-lesson audit.
- Preserved all existing Part I SHA-256 and Part II page/formula/content contracts.

### Exact lessons approved so far — 15 formula IDs

1. `MAI-P2-017`: dot-product prediction and feature contributions.
2. `MAI-P2-020`–`MAI-P2-027`: least squares → MSE → gradient → normal equation → solution.
3. `MAI-P2-028`–`MAI-P2-030`: weighted/unweighted residual gradient and scalar chain rule.
4. `MAI-P2-031`–`MAI-P2-033`: outer-product accumulation and samplewise-vs-stacked least-squares equivalence.

Each group uses a pure numerical core shared by the visual, numeric table, explanation, and invariant checks.

### Validation

PR head `83b8f3d85c6832246decf9e7704a3eaad1095bb7`, Blog CI run `32228530366`: PASS for lint, Astro/Svelte checks, SEO, production build, original source audits, and the new numerical lesson audit.

### Next exact lesson group

`MAI-P2-034`–`MAI-P2-036`: test MSE, expected generalization gap, and ridge/weight-decay trade-off.

The PR remains draft intentionally and will be extended through the remaining Part II groups before merge review.

**Comment by JJo-0** [comment]
## Progress update — generalization and regularization milestone

Completed in draft PR #40:

- [x] `MAI-P2-034` test MSE: selectable test residuals, complete squared-error table, and `SSE/M` invariant.
- [x] `MAI-P2-035` expected train/test error: repeated sample→fit→test process, positive expected gap, and individual split reversals to prevent pointwise misinterpretation.
- [x] `MAI-P2-036` ridge objective: degree-9 source-aligned example, `λ` presets, per-coefficient `w_k²` and `λw_k²`, objective decomposition, and coefficient-shrinkage checks.
- [x] CI `32231659005` green on head `a8bd8fc55fa0f34a5ede5be7aa30733d08a1067e`.

Inventory is now 18 approved formula IDs across 6 exact lesson groups. The next item is K-fold cross-validation; because the source gives a procedure and figure rather than a display formula, it should use a section-level concept registry instead of being forced into an unrelated formula card.

**Comment by JJo-0** [comment]
## 2026-08-20 current-main integration checkpoint

Draft PR #40 has been rebased/ported onto the now-current production architecture after #45 (explicit taxonomy) and #44 (isolated Three.js renderer).

Current #40 head: `887f46229ebc0044dd4c6786797e036142150855`
Blog CI: `32337785671` — GREEN

Preserved contracts now pass together:
- taxonomy: 50 posts / four canonical Research areas
- legacy redirects: 45
- renderer: four canonical Research nodes / 49 GPU-free article routes / SAFE SVG fallback
- Modern AI Part I: 238 protected formulas
- Modern AI Part II: 13 pages / 103 formulas / 65 display-formula walkthroughs
- exact lesson registry: 303 inventoried / 18 approved / six lesson groups / 285 unreviewed with no unrelated fallback visual

The infrastructure/rebase blocker is therefore closed. Remaining work is content/research completion only. Continue in the existing order, beginning with K-fold cross-validation as a section-level concept lesson, then SVM, minimum-distance/prototype geometry, Bayes posterior/evidence, Bayes risk, QDA→LDA→minimum-distance, Naive Bayes/dimensionality growth, followed by the Part I primitives and rendered mobile review before #40 is marked Ready.

**Comment by JJo-0** [comment]
## Part I–II exact-ID formula lesson milestone complete

PR #40 has been squash-merged into `main` as `7e6ffd4ca75b7525fcb434345fde407a88957da9`.

### Completed scope

- fully triaged all **303** protected Part I–II display formulas
  - Part I: **238**
  - Part II: **65**
- **108** exact approved formula IDs across **16** formula-specific lesson groups
- **195** explicit `no-visual-with-reason` decisions
- **0** remaining `unreviewed` formulas
- added one **section-level K-fold cross-validation lesson** without fabricating a formula ID
- completed Part II lessons for SVM/kernel, minimum-distance classification, Bayes posterior/risk, Gaussian QDA→LDA→minimum-distance, Naive Bayes, and dimensionality growth
- completed reusable Part I linear-algebra, eigen/covariance, probability, and optimization primitives
- removed the old generic formula visualization/fallback runtime
- preserved protected formula TeX/SHA-256, Part II PDF page/formula/source-status provenance, including source-suspect vs editorial/corrected variants

### Final validation

Final PR head `f0fa5d8d86999689d0c6a448050cfe068612ab50` passed Blog CI **#277** end to end:

- frozen install / taxonomy / lint: PASS
- Astro + Svelte type diagnostics: PASS
- SEO / production build: PASS
- content integrity + **45 legacy redirects**: PASS
- Part I source-hash/reader contract: PASS (**238 formulas**)
- Part II source contract: PASS (**13 pages / 103 formulas / 65 display walkthroughs / 241 content blocks / 7 figures / 7 annotations**)
- exact formula lesson contract: PASS (**303 inventoried / 108 approved / 195 explicit no-visual / 0 unreviewed / K-fold section lesson / mobile-safe lazy renderers**)
- runtime contract: PASS
- experience / Three.js isolation: PASS (**4 canonical research areas; no article runtime**)
- renderer contract: PASS (**49 GPU-free articles; SAFE SVG/DOM fallback; lazy renderer modules**)
- transition contract: PASS
- rendered-content quality: PASS

### Remaining umbrella scope

Keep #38 **open**. The Part I–II milestone is complete; the remaining work is to apply the same exact-ID / fail-closed contract to Modern AI **III–VIII** as those parts are implemented and published.

**Comment by JJo-0** [comment]
## Modern AI III exact-ID milestone complete — Draft PR #58

Modern AI III has now been migrated to the same exact-formula-ID / fail-closed lesson architecture used by Parts I–II.

### Frozen source boundary

The protected Part III article and ledgers were not modified:

- 15 source pages
- 163 formula records
- 99 display formulas
- 64 inline formulas
- 308 content blocks
- 7 figures
- 11 annotations

### Final Part III lesson partition

- **99 / 99 display formulas approved**
- **10 exact-ID lesson groups**
- **0 unreviewed**
- **0 missing**
- generic fallback renderer: **0**

The ten groups cover perceptron geometry, perceptron learning/error sets, MSE training, XOR geometry, MLP forward/batch/softmax, DropOut/DropConnect, convolution primitives, convolution exercises/separability, boundary/system-matrix forms, and local-patch/CNN/MaxPool formulas.

Source-suspect, source-blank, corrected, and editorial-completion records remain separate; the new audit verifies those provenance links rather than collapsing them.

### Compatibility

The completed Part I–II contract is unchanged:

- 303 protected display formulas
- 108 approved
- 195 explicit `no-visual-with-reason`
- 0 unreviewed
- 16 exact-ID lesson groups
- one section-level K-fold lesson

### Validation

PR #58 final head: `cd9faab8fc4bddfbe59674c8af51c959051e4698`

Blog CI **#392** / run `32448676497`: **GREEN**.

The same run passed frozen install, taxonomy, lint, Astro/Svelte diagnostics, SEO, production build, Part I–III source audits, both exact-lesson audits, formula-guide/runtime contracts, Three.js/article isolation, renderer/transition/rendered-content checks, and all browser smoke jobs. The renderer contract reports **51 GPU-free article routes**.

PR #58 remains **Draft** intentionally. Its implementation/validation is complete; Draft removal and merge are a separate final action.

**Comment by JJo-0** [comment]
## Part III milestone merged

PR #58 was squash-merged to `main` as `10a2209a7847603b06d82f5eb829c1c328f7e074`.

Post-merge verification is content-exact: the merged `main` commit and the CI-validated PR head `cd9faab8fc4bddfbe59674c8af51c959051e4698` both point to tree `703b7ed20d8ac580bf6ff3f43236760cf81c57a6`. Therefore the exact tree validated by Blog CI #392 is what landed on `main`.

Part III remains:

- 99 / 99 display formulas approved
- 10 exact-ID lesson groups
- 0 unreviewed
- 0 missing
- protected Part III source MDX / formula-page-content ledgers unchanged
- Part I–II contract unchanged at 303 = 108 approved + 195 explicit no-visual + 0 unreviewed

Issue #38 stays open for the remaining Parts IV–VIII migration work.

**Comment by JJo-0** [comment]
2026-09-12 사용자 요청에 따른 관리상 정리입니다. 현재 일부 lesson/audit 기반이 있다는 사실과 Modern AI I–VIII의 모든 표시 수식이 exact-ID lesson 계약을 만족한다는 것은 구분해야 합니다. 이번 작업은 수식별 renderer·가정·계산·시각화·숫자 검증과 전체 단계 이행을 재감사하지 않았습니다. 수식 원문과 SHA-256 계약, heuristic 시각화 금지 원칙은 변경하지 않습니다. 원래 체크리스트를 보존하고 `not_planned`로 보관 종료합니다(전체 migration 완료 아님). 잔여 범위는 PR #106 문서에도 기록하며, 재개 시 이 이슈를 다시 열 수 있습니다.

**Comment by JJo-0** [comment]
2026-09-12 사용자 요청에 따라 재개하고 이전 not_planned 정리를 철회합니다. 전체 수식의 exact-ID lesson 연결, 기호·가정·계산·시각화 대응 검증은 계속 진행할 작업입니다. 이번 PR은 Frontier 이미지와 관련한 변경이며 Modern AI 수식이나 SHA 계약을 수정하지 않습니다. 원래 금지한 heuristic 시각화로 대체하거나 수식 개수 일치만으로 완료 처리하지 않습니다. #28의 원자료 매핑과 연결하여 수식군별 단위 시험 및 렌더링 검증 근거가 쌓인 범위만 완료 체크할 예정입니다.
